dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeer.cpp

```diff
@@ -500,19 +593,58 @@ FastAddPeerRes FastAddPeer(EngineStoreServerWrap * server, uint64_t region_id, u
         }
         else
         {
+            const auto & settings = server->tmt->getContext().getSettingsRef();
+            auto maybe_elapsed = fap_ctx->tasks_trace->queryElapsed(region_id);
+            RUNTIME_CHECK_MSG(
+                maybe_elapsed.has_value(),
+                "Task not found, region_id={} new_peer_id={}",
+                region_id,
+                new_peer_id);
+            auto elapsed = maybe_elapsed.value();
+            if (elapsed >= 1000 * settings.fap_task_timeout_seconds)
+            {
+                /// NOTE: Make sure FastAddPeer is the only place to cancel FAP phase 1.
+                // If the task is running, we have to wait it return on cancel and clean,
+                // otherwise a later regular may race with this clean.
+                auto prev_state = fap_ctx->tasks_trace->queryState(region_id);
+                LOG_INFO(
+                    log,
+                    "Cancel FAP due to timeout region_id={} new_peer_id={} prev_state={}",
+                    region_id,
+                    new_peer_id,
+                    magic_enum::enum_name(prev_state));
+                GET_METRIC(tiflash_fap_task_state, type_blocking_cancel_stage).Increment();
+                if (prev_state == FAPAsyncTasks::TaskState::Running)
+                {
+                    [[maybe_unused]] auto s = fap_ctx->tasks_trace->blockedCancelRunningTask(region_id);
+                }
+                GET_METRIC(tiflash_fap_task_state, type_blocking_cancel_stage).Decrement();
```

Comment:
Now you only cal `blockedCancelRunningTask` when prev_state == FAPAsyncTasks::TaskState::Running, it is not the same logic as the commit before. Is it expected?

Reply:
No, it is a typo, actually the test fires locally... I will have a push later